src/codegen/mod.rs

```diff
@@ -0,0 +1,2000 @@
+use ir::int::IntKind;
+use ir::module::Module;
+use ir::var::Var;
+use ir::enum_ty::{Enum, EnumVariant};
+use ir::function::{Function, FunctionSig};
+use ir::item_kind::ItemKind;
+use ir::comp::{CompKind, CompInfo, Field, Method};
+use ir::layout::Layout;
+use ir::annotations::FieldAccessorKind;
+
+use std::ops;
+use std::mem;
+use std::collections::BTreeSet;
+use std::collections::HashSet;
+use std::collections::hash_map::{HashMap, Entry};
+
+use syntax::abi::Abi;
+use syntax::ast;
+use syntax::codemap::{Span, respan};
+use syntax::ptr::P;
+use aster;
+
+fn root_import(ctx: &BindgenContext) -> P<ast::Item> {
+    assert!(ctx.options().enable_cxx_namespaces, "Somebody messed it up");
+    let root = ctx.root_module().canonical_name(ctx);
+    let root_ident = ctx.rust_ident(&root);
+    quote_item!(ctx.ext_cx(), use $root_ident;).unwrap()
+}
+
+macro_rules! repr {
+    ($which:expr) => {{
+        aster::AstBuilder::new().attr().list("repr").words(&[$which]).build()
+    }}
+}
+
+macro_rules! doc {
+    ($comment:expr) => {{
+        aster::AstBuilder::new().attr().doc($comment)
+    }}
+}
+
+macro_rules! link_name {
+    ($name:expr) => {{
+        aster::AstBuilder::new().attr().name_value("link_name").str($name)
+    }}
+}
+
+struct CodegenResult {
+    items: Vec<P<ast::Item>>,
+    saw_union: bool,
+    items_seen: HashSet<ItemId>,
+    /// The set of generated function names, needed because in C/C++ is legal to
+    /// do something like:
+    ///
+    /// ```
+    /// extern "C" {
+    ///   void foo();
+    /// }
+    ///
+    /// extern "C" {
+    ///   void foo();
+    /// }
+    /// ```
+    ///
+    /// Being these two different declarations.
+    functions_seen: HashSet<String>,
+}
+
+impl CodegenResult {
+    fn new() -> Self {
+        CodegenResult {
+            items: vec![],
+            saw_union: false,
+            items_seen: Default::default(),
+            functions_seen: Default::default(),
+        }
+    }
+
+    fn saw_union(&mut self) {
+        self.saw_union = true;
+    }
+
+    fn seen(&self, item: ItemId) -> bool {
+        self.items_seen.contains(&item)
+    }
+
+    fn set_seen(&mut self, item: ItemId) {
+        self.items_seen.insert(item);
+    }
+
+    fn seen_function(&self, name: &String) -> bool {
+        self.functions_seen.contains(name)
+    }
+
+    fn saw_function(&mut self, name: &str) {
+        self.functions_seen.insert(name.into());
+    }
+
+    fn inner<F>(&mut self, cb: F) -> Vec<P<ast::Item>>
+        where F: FnOnce(&mut Self)
+    {
+        let mut new = Self::new();
+
+        cb(&mut new);
+
+        self.saw_union |= new.saw_union;
+
+        new.items
+    }
+}
+
+impl ops::Deref for CodegenResult {
+    type Target = Vec<P<ast::Item>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.items
+    }
+}
+
+impl ops::DerefMut for CodegenResult {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.items
+    }
+}
+
+struct ForeignModBuilder {
```

Comment:
Sure, I'll move most of this to a `codegen/builders`, `codegen/helpers`, or similar.